### PR TITLE
[processing][needs-docs] use native syntax for Processing scripts

### PR DESCRIPTION
**Pull request by alexbruy**
## Description
Drop old Processing scripts syntax, use native Python syntax instead.

Each script algorithm should de defined in the separate file. Script algorithms are subclasses of the `QgsProcessingAlgorithm` class and should implement all required methods as any other Processing algorithm.

Related discussion in mailing list https://lists.osgeo.org/pipermail/qgis-developer/2018-January/051511.html

No new unit tests, as `QgsProcessingAlgorithm` already covered by tests.

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [x] Commit messages are descriptive and explain the rationale for changes
- [ ] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [ ] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [x] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and containt sufficient information in the commit message to be documented
- [x] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [x] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [ ] New unit tests have been added for core changes
- [x] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTE.md#contributing-to-qgis) before each commit

### CODE REVIEW

**Comment by nyalldawson** [comment]
> In addition it is necessary to define __all__ variable in the script source file like where "SpatialIndex" is the name of the algorithm class. This is necessary to allow saving scripts in the file with any name. Alternatively we can require saving scripts in the files with the same name as algorithm class name and use a bit of introspection magic to load required class without knowing its name in advance.

Thinking about this... would it be possible instead to introspect every class defined in the script file, and if any are derived from QgsProcessingAlgorithm THEN they get added as an algorithm?

**Comment by nyalldawson** [comment]
>..."as QgsCodeEditor is not available on some platforms"

Is that still the case? Or just an outdated comment?

**Comment by alexbruy** [comment]
> > "as QgsCodeEditor is not available on some platforms"
>
> Is that still the case? Or just an outdated comment?

Unfortunately, seems it is still valid. At least on Travis corresponding import failed. I can try again switch to QgsCodeEditor and check.

**Comment by alexbruy** [comment]
> would it be possible instead to introspect every class defined in the script file, and if any are derived from QgsProcessingAlgorithm THEN they get added as an algorithm?

Yes, this was my inital approach, will use it then.

**Comment by alexbruy** [comment]
@nyalldawson now algorithm class loaded using introspection, no need to specify class name explicitly. Still one algorithm per file, but it can be easily adopted to allow multiple algorithms per file

**Comment by alexbruy** [comment]
Now trying to switch back to `QgsCodeEditor`.

**Comment by alexbruy** [comment]
@nyalldawson, Travis failed with
```
ImportError: cannot import name 'QgsCodeEditorPython'
```
when using `QgsCodeEditor`, looks like it still not available on some platforms.

**Comment by alexbruy** [comment]
Also removed obsolete code from core files.

File: python/plugins/processing/script/ScriptEditorDialog.py [hunk 1]
```diff
@@ -194,10 +197,28 @@ def setHasChanged(self, hasChanged):
...
+        except Exception as e:
+            error = QgsError(traceback.format_exc(), "Processing")
+            QgsErrorDialog.show(error,
+                                self.tr("Execution error")
+                                )
+            return
+
+        alg = None
+        for k, v in d.items():
+            if inspect.isclass(v) and issubclass(v, QgsProcessingAlgorithm) and v.__name__ != "QgsProcessingAlgorithm":
```
Comment (nyalldawson) [inline python/plugins/processing/script/ScriptEditorDialog.py]: I wonder if we need to add "QgsProcessingFeatureBasedAlgorithm" to the name check here - I can imagine cases where a script algorithm could/should be based off QgsProcessingFeatureBasedAlgorithm also.

File: src/core/processing/models/qgsprocessingmodelalgorithm.cpp [hunk 2]
```diff
@@ -334,92 +334,6 @@ void QgsProcessingModelAlgorithm::setSourceFilePath( const QString &sourceFile )
   mSourceFile = sourceFile;
 }
 
-QString QgsProcessingModelAlgorithm::asPythonCode() const
```
Comment (nyalldawson) [inline src/core/processing/models/qgsprocessingmodelalgorithm.cpp]: Hmm - we're gonna get flack for removing this. Can we just ifdef it out for now, and then someone can potentially upgrade it to create proper algorithms at some later stage?

File: python/plugins/processing/script/ScriptEditorDialog.py (hunk 1 above)
Comment (alexbruy) [reply python/plugins/processing/script/ScriptEditorDialog.py]: I also thought about adding feature-based alg here. Will add it now.

**Comment by alexbruy** [comment]
@nyalldawson comments addressed now.

**Comment by alexbruy** [comment]
@nyalldawson what is your opinion, can we merge this before release?

**Comment by nyalldawson** [comment]
This is a strong +1 from me. I think we should merge and then get feedback from the dev list?

**Pull request merged**